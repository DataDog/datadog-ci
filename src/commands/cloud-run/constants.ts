import {ENVIRONMENT_ENV_VAR, SERVICE_ENV_VAR, SITE_ENV_VAR} from '../../constants'

const NODE_OPTIONS_ENV_VAR = 'NODE_OPTIONS'
const DD_TRACE_PROPAGATION_STYLE_ENV_VAR = 'DD_TRACE_PROPAGATION_STYLE'
const DD_SOURCE_ENV_VAR = 'DD_SOURCE'
const DD_TAGS_ENV_VAR = 'DD_TAGS'
const DD_OLTP_HTTPS_ENV_VAR = 'DD_OTLP_CONFIG_RECEIVER_PROTOCOLS_HTTP_ENDPOINT'
const DD_OLTP_GRPC_ENV_VAR = 'DD_OTLP_CONFIG_RECEIVER_PROTOCOLS_GRPC_ENDPOINT'

export const SKIP_MASKING_CLOUDRUN_ENV_VARS = new Set([
  SITE_ENV_VAR,
  SERVICE_ENV_VAR,
  ENVIRONMENT_ENV_VAR,
  NODE_OPTIONS_ENV_VAR,
  DD_TRACE_PROPAGATION_STYLE_ENV_VAR,
  DD_SOURCE_ENV_VAR,
  DD_TAGS_ENV_VAR,
  DD_OLTP_HTTPS_ENV_VAR,
  DD_OLTP_GRPC_ENV_VAR,
])
