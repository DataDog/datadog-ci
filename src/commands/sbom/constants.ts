export const API_ENDPOINT = 'api/v2/static-analysis-sca/dependencies'
