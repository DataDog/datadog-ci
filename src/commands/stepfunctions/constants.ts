export const DD_CI_IDENTIFYING_STRING = 'DdCiLogGroupSubscription'
export const DD_TRACE_ENABLED = 'DD_TRACE_ENABLED'
export const TAG_VERSION_NAME = 'dd_sls_ci'
