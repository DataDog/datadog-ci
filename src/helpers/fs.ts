/**
 * @file Functions that interact with the file system.
 * Functions have built-in error handling for more descriptive error messages.
 */

import fs from 'fs'
import path from 'path'

import JSZip from 'jszip'

/**
 * Delete a folder and all its contents
 * @param folderPath the folder to delete
 * @throws Error if the deletion fails
 */
export const deleteFolder = (folderPath: string) => {
  try {
    fs.rmSync(folderPath, {recursive: true, force: true})
  } catch (err) {
    if (err instanceof Error) {
      throw Error(`Failed to delete files located at ${folderPath}: ${err.message}`)
    }
  }
}

/**
 * Write the data to a file
 * @param filePath path to the file
 * @param data the data to write
 * @throws Error if the file cannot be written
 */
export const writeFile = (filePath: string, data: string) => {
  try {
    fs.writeFileSync(filePath, data)
  } catch (err) {
    if (err instanceof Error) {
      throw Error(`Unable to write file: ${err.message}`)
    }
  }
}

/**
 * Zip the entire contents of a folder
 * @param rootFolderPath path to the root folder to zip
 * @param zipPath path to save the zip file
 * @throws Error if the zip fails
 */
export const zipContents = async (rootFolderPath: string, zipPath: string) => {
  const zip = new JSZip()

  const addFolderToZip = (folderPath: string) => {
    if (!fs.existsSync(folderPath)) {
      throw Error(`Folder does not exist: ${folderPath}`)
    }

    const folder = fs.statSync(folderPath)
    if (!folder.isDirectory()) {
      throw Error(`Path is not a directory: ${folderPath}`)
    }

    const contents = fs.readdirSync(folderPath)
    for (const item of contents) {
      const fullPath = path.join(folderPath, item)
      const file = fs.statSync(fullPath)

      if (file.isDirectory()) {
        addFolderToZip(fullPath)
      } else {
        const data = fs.readFileSync(fullPath)
        zip.file(path.relative(rootFolderPath, fullPath), data)
      }
    }
  }

  try {
    addFolderToZip(rootFolderPath)
    const zipContent = await zip.generateAsync({type: 'nodebuffer'})
    fs.writeFileSync(zipPath, zipContent)
  } catch (err) {
    if (err instanceof Error) {
      throw Error(`Unable to zip files: ${err.message}`)
    }
  }
}

/**
 * Creates the root folder and any subfolders
 * @param rootFolderPath path to the root folder
 * @param subFolders paths to any subfolders to be created
 * @throws Error if the root folder cannot be deleted or folders cannot be created
 */
export const createDirectories = (rootFolderPath: string, subFolders: string[]) => {
  try {
    fs.mkdirSync(rootFolderPath)
    for (const subFolder of subFolders) {
      fs.mkdirSync(subFolder)
    }
  } catch (err) {
    if (err instanceof Error) {
      throw Error(`Unable to create directories: ${err.message}`)
    }
  }
}
