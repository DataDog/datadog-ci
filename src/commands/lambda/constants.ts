import {ENVIRONMENT_ENV_VAR, FLARE_PROJECT_FILES, SERVICE_ENV_VAR, SITE_ENV_VAR, VERSION_ENV_VAR} from '../../constants'

export const DD_LAMBDA_EXTENSION_LAYER_NAME = 'Datadog-Extension'
export const EXTENSION_LAYER_KEY = 'extension'
export const LAYER_LOOKUP = {
  [EXTENSION_LAYER_KEY]: DD_LAMBDA_EXTENSION_LAYER_NAME,
  dotnet6: 'dd-trace-dotnet',
  'dotnetcore3.1': 'dd-trace-dotnet',
  java11: 'dd-trace-java',
  java17: 'dd-trace-java',
  'java8.al2': 'dd-trace-java',
  'nodejs12.x': 'Datadog-Node12-x',
  'nodejs14.x': 'Datadog-Node14-x',
  'nodejs16.x': 'Datadog-Node16-x',
  'nodejs18.x': 'Datadog-Node18-x',
  'python3.7': 'Datadog-Python37',
  'python3.8': 'Datadog-Python38',
  'python3.9': 'Datadog-Python39',
  'python3.10': 'Datadog-Python310',
  'python3.11': 'Datadog-Python311',
  'ruby2.7': 'Datadog-Ruby2-7',
  'ruby3.2': 'Datadog-Ruby3-2',
} as const

export enum RuntimeType {
  DOTNET,
  CUSTOM,
  JAVA,
  NODE,
  PYTHON,
  RUBY,
}

export const RUNTIME_LOOKUP = {
  dotnet6: RuntimeType.DOTNET,
  'dotnetcore3.1': RuntimeType.DOTNET,
  java11: RuntimeType.JAVA,
  java17: RuntimeType.JAVA,
  'java8.al2': RuntimeType.JAVA,
  'nodejs12.x': RuntimeType.NODE,
  'nodejs14.x': RuntimeType.NODE,
  'nodejs16.x': RuntimeType.NODE,
  'nodejs18.x': RuntimeType.NODE,
  'provided.al2': RuntimeType.CUSTOM,
  'python3.7': RuntimeType.PYTHON,
  'python3.8': RuntimeType.PYTHON,
  'python3.9': RuntimeType.PYTHON,
  'python3.10': RuntimeType.PYTHON,
  'python3.11': RuntimeType.PYTHON,
  'ruby2.7': RuntimeType.RUBY,
  'ruby3.2': RuntimeType.RUBY,
}

export type Runtime = keyof typeof RUNTIME_LOOKUP
export type LayerKey = keyof typeof LAYER_LOOKUP
export const ARM_LAYERS = [
  EXTENSION_LAYER_KEY,
  'dotnet6',
  'python3.8',
  'python3.9',
  'python3.10',
  'python3.11',
  'ruby2.7',
  'ruby3.2',
]
export const ARM64_ARCHITECTURE = 'arm64'
export const ARM_LAYER_SUFFIX = '-ARM'

export const PYTHON_HANDLER_LOCATION = 'datadog_lambda.handler.handler'
export const NODE_HANDLER_LOCATION = '/opt/nodejs/node_modules/datadog-lambda-js/handler.handler'

export const DEFAULT_LAYER_AWS_ACCOUNT = '464622532012'
export const GOVCLOUD_LAYER_AWS_ACCOUNT = '002406178527'
export const SUBSCRIPTION_FILTER_NAME = 'datadog-ci-filter'
export const TAG_VERSION_NAME = 'dd_sls_ci'

// Env variables for Univeral instrument lambda exec wrapper
export const AWS_LAMBDA_EXEC_WRAPPER_VAR = 'AWS_LAMBDA_EXEC_WRAPPER'
export const AWS_LAMBDA_EXEC_WRAPPER = '/opt/datadog_wrapper'

// Export const values for .NET tracer
export const CORECLR_ENABLE_PROFILING = '1'
export const CORECLR_PROFILER = '{846F5F1C-F9AE-4B07-969E-05C26BC060D8}'
export const CORECLR_PROFILER_PATH = '/opt/datadog/Datadog.Trace.ClrProfiler.Native.so'
export const DD_DOTNET_TRACER_HOME = '/opt/datadog'

// Environment variables used in the Lambda environment
export const API_KEY_SECRET_ARN_ENV_VAR = 'DD_API_KEY_SECRET_ARN'
export const KMS_API_KEY_ENV_VAR = 'DD_KMS_API_KEY'
export const TRACE_ENABLED_ENV_VAR = 'DD_TRACE_ENABLED'
export const MERGE_XRAY_TRACES_ENV_VAR = 'DD_MERGE_XRAY_TRACES'
export const FLUSH_TO_LOG_ENV_VAR = 'DD_FLUSH_TO_LOG'
export const LOG_LEVEL_ENV_VAR = 'DD_LOG_LEVEL'
export const LAMBDA_HANDLER_ENV_VAR = 'DD_LAMBDA_HANDLER'
export const EXTRA_TAGS_ENV_VAR = 'DD_TAGS'
export const CAPTURE_LAMBDA_PAYLOAD_ENV_VAR = 'DD_CAPTURE_LAMBDA_PAYLOAD'
export const APM_FLUSH_DEADLINE_MILLISECONDS_ENV_VAR = 'DD_APM_FLUSH_DEADLINE_MILLISECONDS'
export const APPSEC_ENABLED_ENV_VAR = 'DD_SERVERLESS_APPSEC_ENABLED'
export const ENABLE_PROFILING_ENV_VAR = 'CORECLR_ENABLE_PROFILING'
export const PROFILER_ENV_VAR = 'CORECLR_PROFILER'
export const PROFILER_PATH_ENV_VAR = 'CORECLR_PROFILER_PATH'
export const DOTNET_TRACER_HOME_ENV_VAR = 'DD_DOTNET_TRACER_HOME'

// Environment variables used by Datadog CI
export const CI_API_KEY_SECRET_ARN_ENV_VAR = 'DATADOG_API_KEY_SECRET_ARN'
export const CI_KMS_API_KEY_ENV_VAR = 'DATADOG_KMS_API_KEY'

export const AWS_ACCESS_KEY_ID_ENV_VAR = 'AWS_ACCESS_KEY_ID'
export const AWS_SECRET_ACCESS_KEY_ENV_VAR = 'AWS_SECRET_ACCESS_KEY'
export const AWS_DEFAULT_REGION_ENV_VAR = 'AWS_DEFAULT_REGION'
export const AWS_SESSION_TOKEN_ENV_VAR = 'AWS_SESSION_TOKEN'
export const AWS_SHARED_CREDENTIALS_FILE_ENV_VAR = 'AWS_SHARED_CREDENTIALS_FILE'

export const LIST_FUNCTIONS_MAX_RETRY_COUNT = 2
export const MAX_LAMBDA_STATE_CHECK_ATTEMPTS = 3

// DD_TAGS Regular Expression
// This RegExp ensures that the --extra-tags string
// matches a list of <key>:<value> separated by commas
// such as: layer:api,team:intake
export const EXTRA_TAGS_REG_EXP = /^(([a-zA-Z]+)[\w\-/\.]*:[^,]+)+((\,)([a-zA-Z]+)[\w\-/\.]*:[^,]+)*$/g
export const AWS_ACCESS_KEY_ID_REG_EXP = /(?<![A-Z0-9])[A-Z0-9]{20}(?![A-Z0-9])/g
export const AWS_SECRET_ACCESS_KEY_REG_EXP = /(?<![A-Za-z0-9/+=])[A-Za-z0-9/+=]{40}(?![A-Za-z0-9/+=])/g
export const AWS_SECRET_ARN_REG_EXP = /arn:aws:secretsmanager:[\w-]+:\d{12}:secret:.+/
export const DATADOG_API_KEY_REG_EXP = /(?<![a-f0-9])[a-f0-9]{32}(?![a-f0-9])/g
export const DATADOG_APP_KEY_REG_EXP = /(?<![a-f0-9])[a-f0-9]{40}(?![a-f0-9])/g

// Environment Variables whose values don't need to be masked
export const SKIP_MASKING_LAMBDA_ENV_VARS = new Set([
  AWS_LAMBDA_EXEC_WRAPPER_VAR,
  API_KEY_SECRET_ARN_ENV_VAR,
  DOTNET_TRACER_HOME_ENV_VAR,
  ENVIRONMENT_ENV_VAR,
  EXTRA_TAGS_ENV_VAR,
  LAMBDA_HANDLER_ENV_VAR,
  LOG_LEVEL_ENV_VAR,
  KMS_API_KEY_ENV_VAR,
  PROFILER_ENV_VAR,
  PROFILER_PATH_ENV_VAR,
  SERVICE_ENV_VAR,
  SITE_ENV_VAR,
  VERSION_ENV_VAR,
])

export enum DeploymentFrameworks {
  ServerlessFramework = 'Serverless Framework',
  AwsCdk = 'AWS CDK',
  AwsCloudFormation = 'AWS CloudFormation',
  Unknown = 'Unknown',
}

// Mappings of files to frameworks.
// For example, if `serverless.yml` exists, we know it's the Serverless Framework
export const FRAMEWORK_FILES_MAPPING = new Map([
  ['serverless.yaml', DeploymentFrameworks.ServerlessFramework],
  ['serverless.yml', DeploymentFrameworks.ServerlessFramework],
  ['serverless.js', DeploymentFrameworks.ServerlessFramework],
  ['cdk.json', DeploymentFrameworks.AwsCdk],
  ['.cdk.json', DeploymentFrameworks.AwsCdk],
  ['template.yaml', DeploymentFrameworks.AwsCloudFormation],
  ['template.yml', DeploymentFrameworks.AwsCloudFormation],
  ['template.json', DeploymentFrameworks.AwsCloudFormation],
])

export const LAMBDA_PROJECT_FILES = [...FLARE_PROJECT_FILES, ...FRAMEWORK_FILES_MAPPING.keys()]
