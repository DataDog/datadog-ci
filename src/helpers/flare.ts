/**
 * @file Functions used by `lambda flare` and `cloud-run flare`.
 */

import fs from 'fs'
import path from 'path'

import axios from 'axios'
import FormData from 'form-data'

import {
  CI_SITE_ENV_VAR,
  DATADOG_SITE_EU1,
  DATADOG_SITE_GOV,
  DATADOG_SITE_US1,
  DATADOG_SITES,
  FLARE_ENDPOINT_PATH,
  SITE_ENV_VAR,
} from '../constants'

import {deleteFolder} from './fs'
import * as helpersRenderer from './renderer'
import {isValidDatadogSite} from './validation'
import {version} from './version'

/**
 * Send the zip file to Datadog support
 * @param zipPath
 * @param caseId
 * @param email
 * @param apiKey
 * @param rootFolderPath
 * @throws Error if the request fails
 */
export const sendToDatadog = async (
  zipPath: string,
  caseId: string,
  email: string,
  apiKey: string,
  rootFolderPath: string
) => {
  const endpointUrl = getEndpointUrl()
  const form = new FormData()
  form.append('case_id', caseId)
  form.append('flare_file', fs.createReadStream(zipPath))
  form.append('datadog_ci_version', version)
  form.append('email', email)
  const headerConfig = {
    headers: {
      ...form.getHeaders(),
      'DD-API-KEY': apiKey,
    },
  }

  try {
    await axios.post(endpointUrl, form, headerConfig)
  } catch (err) {
    // Ensure the root folder is deleted if the request fails
    deleteFolder(rootFolderPath)

    if (axios.isAxiosError(err)) {
      const errResponse: string = (err.response?.data.error as string) ?? ''
      const errorMessage = err.message ?? ''

      let message = `Failed to send flare file to Datadog Support: ${errorMessage}. ${errResponse}\n`
      const code = err.response?.status
      // The error message doesn't say why there was an error. All it says is:
      // "[Error] Failed to send flare file to Datadog Support: Request failed with status code 500."
      // Therefore, we need to add an explanation to clarify when the code is 500 or 403.
      switch (code) {
        case 500:
          message += 'Are your case ID and email correct?\n'
          break
        case 400:
        case 403:
          message += 'Is your Datadog API key correct?\n'
          break
      }

      throw Error(message)
    }

    throw err
  }
}

/**
 * Calculates the full endpoint URL
 * @throws Error if the site is invalid
 * @returns the full endpoint URL
 */
export const getEndpointUrl = () => {
  const baseUrl = process.env[CI_SITE_ENV_VAR] ?? process.env[SITE_ENV_VAR] ?? DATADOG_SITE_US1
  // The DNS doesn't redirect to the proper endpoint when a subdomain is not present in the baseUrl.
  // There is a DNS inconsistency
  let endpointUrl = baseUrl
  if ([DATADOG_SITE_US1, DATADOG_SITE_EU1, DATADOG_SITE_GOV].includes(baseUrl)) {
    endpointUrl = 'app.' + baseUrl
  }

  if (!isValidDatadogSite(baseUrl)) {
    throw Error(`Invalid site: ${baseUrl}. Must be one of: ${DATADOG_SITES.join(', ')}`)
  }

  return 'https://' + endpointUrl + FLARE_ENDPOINT_PATH
}

/**
 * Searches current directory for project files
 * @returns a set of file paths of project files
 */
export const getProjectFiles = async (projectFiles: string[]) => {
  const filePaths = new Set<string>()
  const cwd = process.cwd()
  for (const fileName of projectFiles) {
    const filePath = path.join(cwd, fileName)
    if (fs.existsSync(filePath)) {
      filePaths.add(filePath)
    }
  }

  return filePaths
}

/**
 * Validates a path to a file
 * @param filePath path to the file
 * @param projectFilePaths map of file names to file paths
 * @param additionalFiles set of additional file paths
 * @throws Error if the file path is invalid or the file was already added
 * @returns the full path to the file
 */
export const validateFilePath = (filePath: string, projectFilePaths: Set<string>, additionalFiles: Set<string>) => {
  const originalPath = filePath
  filePath = fs.existsSync(filePath) ? filePath : path.join(process.cwd(), filePath)
  if (!fs.existsSync(filePath)) {
    throw Error(helpersRenderer.renderError(`File path '${originalPath}' not found. Please try again.`))
  }

  filePath = path.resolve(filePath)
  if (projectFilePaths.has(filePath) || additionalFiles.has(filePath)) {
    throw Error(helpersRenderer.renderSoftWarning(`File '${filePath}' has already been added.`))
  }

  return filePath
}

/**
 * Validate the start and end flags and adds error messages if found
 * @param start start time as a string
 * @param end end time as a string
 * @throws error if start or end are not valid numbers
 * @returns [startMillis, endMillis] as numbers or [undefined, undefined] if both are undefined
 */
export const validateStartEndFlags = (start: string | undefined, end: string | undefined) => {
  if (!start && !end) {
    return [undefined, undefined]
  }

  if (!start) {
    throw Error('Start time is required when end time is specified. [--start]')
  }
  if (!end) {
    throw Error('End time is required when start time is specified. [--end]')
  }

  let startMillis = Number(start)
  let endMillis = Number(end)
  if (isNaN(startMillis)) {
    throw Error(`Start time must be a time in milliseconds since Unix Epoch. '${start}' is not a number.`)
  }
  if (isNaN(endMillis)) {
    throw Error(`End time must be a time in milliseconds since Unix Epoch. '${end}' is not a number.`)
  }

  // Required for AWS SDK to work correctly
  startMillis = Math.min(startMillis, Date.now())
  endMillis = Math.min(endMillis, Date.now())

  if (startMillis >= endMillis) {
    throw Error('Start time must be before end time.')
  }

  return [startMillis, endMillis]
}
