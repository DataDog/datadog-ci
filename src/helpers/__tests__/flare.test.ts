import fs from 'fs'
import process from 'process'

import axios from 'axios'
import FormData from 'form-data'

import {CI_SITE_ENV_VAR, FLARE_PROJECT_FILES, SITE_ENV_VAR} from '../../constants'

import {getEndpointUrl, getProjectFiles, sendToDatadog, validateFilePath, validateStartEndFlags} from '../flare'
import * as flareModule from '../flare'

import {MOCK_CWD} from './fixtures'

// Mocks
jest.mock('fs')
jest.spyOn(process, 'cwd').mockReturnValue(MOCK_CWD)
jest.spyOn(flareModule, 'getProjectFiles').mockResolvedValue(new Set())
fs.createReadStream = jest.fn().mockReturnValue('test data')
jest.mock('jszip')

describe('flare', () => {
  describe('getEndpointUrl', () => {
    const ORIGINAL_ENV = process.env

    beforeEach(() => {
      process.env = {...ORIGINAL_ENV}
    })

    afterAll(() => {
      process.env = ORIGINAL_ENV
    })

    it('should return correct endpoint url', () => {
      process.env[CI_SITE_ENV_VAR] = 'datadoghq.com'
      const url = getEndpointUrl()
      expect(url).toMatchSnapshot()
    })

    it('should throw error if the site is invalid', () => {
      process.env[CI_SITE_ENV_VAR] = 'datad0ge.com'
      expect(() => getEndpointUrl()).toThrowErrorMatchingSnapshot()
    })

    it('should not throw error if the site is invalid and DD_CI_BYPASS_SITE_VALIDATION is set', () => {
      process.env['DD_CI_BYPASS_SITE_VALIDATION'] = 'true'
      process.env[CI_SITE_ENV_VAR] = 'datad0ge.com'
      const url = getEndpointUrl()
      expect(url).toMatchSnapshot()
      delete process.env['DD_CI_BYPASS_SITE_VALIDATION']
    })

    it('should use SITE_ENV_VAR if CI_SITE_ENV_VAR is not set', () => {
      delete process.env[CI_SITE_ENV_VAR]
      process.env[SITE_ENV_VAR] = 'us3.datadoghq.com'
      const url = getEndpointUrl()
      expect(url).toMatchSnapshot()
    })

    it('should use DEFAULT_DD_SITE if CI_SITE_ENV_VAR and SITE_ENV_VAR are not set', () => {
      delete process.env[CI_SITE_ENV_VAR]
      delete process.env[SITE_ENV_VAR]
      const url = getEndpointUrl()
      expect(url).toMatchSnapshot()
    })
  })

  describe('sendToDatadog', () => {
    const MOCK_ZIP_PATH = '/path/to/zip'
    const MOCK_CASE_ID = 'case1234'
    const MOCK_EMAIL = 'test@example.com'
    const MOCK_API_KEY = 'api-key'
    const MOCK_ROOT_FOLDER_PATH = '/root/folder/path'
    const MOCK_AXIOS = axios as jest.Mocked<typeof axios>

    it('should send data to the correct endpoint', async () => {
      await sendToDatadog(MOCK_ZIP_PATH, MOCK_CASE_ID, MOCK_EMAIL, MOCK_API_KEY, MOCK_ROOT_FOLDER_PATH)
      expect(MOCK_AXIOS.post).toHaveBeenCalledWith(
        expect.any(String),
        expect.any(FormData),
        expect.objectContaining({
          headers: expect.objectContaining({
            'DD-API-KEY': MOCK_API_KEY,
          }),
        })
      )
    })

    it('should delete root folder and rethrow error if request fails', async () => {
      const error = new Error('Network error')
      MOCK_AXIOS.post.mockRejectedValueOnce({
        isAxiosError: true,
        message: error.message,
        response: {data: {error: 'Server error'}},
      })

      const fn = sendToDatadog(MOCK_ZIP_PATH, MOCK_CASE_ID, MOCK_EMAIL, MOCK_API_KEY, MOCK_ROOT_FOLDER_PATH)
      await expect(fn).rejects.toThrow(`Failed to send flare file to Datadog Support: ${error.message}. Server error\n`)
    })

    it('prints correct warning when post fail with error 500', async () => {
      MOCK_AXIOS.post.mockRejectedValueOnce({
        isAxiosError: true,
        message: 'Some error',
        response: {status: 500, data: {error: 'Server error'}},
      })

      const fn = sendToDatadog(MOCK_ZIP_PATH, MOCK_CASE_ID, MOCK_EMAIL, MOCK_API_KEY, MOCK_ROOT_FOLDER_PATH)
      await expect(fn).rejects.toThrow(
        `Failed to send flare file to Datadog Support: Some error. Server error\nAre your case ID and email correct?\n`
      )
    })

    it('prints correct warning when post fail with error 403', async () => {
      MOCK_AXIOS.post.mockRejectedValueOnce({
        isAxiosError: true,
        message: 'Some error',
        response: {status: 403, data: {error: 'Another error'}},
      })

      const fn = sendToDatadog(MOCK_ZIP_PATH, MOCK_CASE_ID, MOCK_EMAIL, MOCK_API_KEY, MOCK_ROOT_FOLDER_PATH)
      await expect(fn).rejects.toThrow(
        `Failed to send flare file to Datadog Support: Some error. Another error\nIs your Datadog API key correct?\n`
      )
    })
  })

  describe('getProjectFiles', () => {
    beforeAll(() => {
      ;(flareModule.getProjectFiles as jest.Mock).mockRestore()
      ;(process.cwd as jest.Mock).mockReturnValue('')
    })

    it('should return a map of existing project files', async () => {
      const mockFiles = ['serverless.yml', 'package.json']
      ;(fs.existsSync as jest.Mock).mockImplementation((filePath: string) => mockFiles.includes(filePath))

      const result = await getProjectFiles(FLARE_PROJECT_FILES)
      expect(Array.from(result.keys())).toEqual(['package.json'])
      expect(fs.existsSync).toHaveBeenCalledTimes(FLARE_PROJECT_FILES.length)
    })

    it('should return an empty map when no files exist', async () => {
      ;(fs.existsSync as jest.Mock).mockReturnValue(false)

      const result = await getProjectFiles(FLARE_PROJECT_FILES)
      expect(result).toEqual(new Set())
      expect(fs.existsSync).toHaveBeenCalledTimes(FLARE_PROJECT_FILES.length)
    })
  })

  describe('validateFilePath', () => {
    const projectFilePaths = new Set<string>()
    const additionalFilePaths = new Set<string>()

    it('returns the correct path when the file exists', () => {
      const filePath = '/exists'

      ;(fs.existsSync as jest.Mock).mockReturnValue(true)
      const result = validateFilePath(filePath, projectFilePaths, additionalFilePaths)

      expect(result).toBe(filePath)
      expect(fs.existsSync).toHaveBeenCalledWith(filePath)
    })

    it('returns the correct path when the file exists relative to the cwd', () => {
      const filePath = 'relative'

      ;(fs.existsSync as jest.Mock).mockReturnValueOnce(false).mockReturnValueOnce(true)

      const result = validateFilePath(filePath, projectFilePaths, additionalFilePaths)

      expect(result).toContain(filePath)
      expect(fs.existsSync).toHaveBeenNthCalledWith(1, filePath)
      expect(fs.existsSync).toHaveBeenCalledTimes(2)
    })

    it('throws an error when the file does not exist', async () => {
      const filePath = '/not-exists'

      ;(fs.existsSync as jest.Mock).mockReturnValue(false)

      expect(() => validateFilePath(filePath, projectFilePaths, additionalFilePaths)).toThrowErrorMatchingSnapshot()
      expect(fs.existsSync).toHaveBeenCalledWith(filePath)
    })

    it('throws an error when the file has already been added', async () => {
      const filePath = '/added'

      ;(fs.existsSync as jest.Mock).mockReturnValue(true)
      projectFilePaths.add(filePath)

      expect(() => validateFilePath(filePath, projectFilePaths, additionalFilePaths)).toThrowErrorMatchingSnapshot()
      expect(fs.existsSync).toHaveBeenCalledWith(filePath)
    })
  })

  describe('validateStartEndFlags', () => {
    beforeEach(() => {
      jest.useFakeTimers({now: new Date(Date.UTC(2023, 0))})
    })

    afterEach(() => {
      jest.useRealTimers()
    })

    it('returns [undefined, undefined] when start and end flags are not specified', () => {
      const errorMessages: string[] = []
      const res = validateStartEndFlags(undefined, undefined)
      expect(res).toEqual([undefined, undefined])
      expect(errorMessages).toEqual([])
    })

    it('throws error when start is specified but end is not specified', () => {
      expect(() => validateStartEndFlags('123', undefined)).toThrowErrorMatchingSnapshot()
    })

    it('throws error when end is specified but start is not specified', () => {
      expect(() => validateStartEndFlags(undefined, '123')).toThrowErrorMatchingSnapshot()
    })

    it('throws error when start is invalid', () => {
      expect(() => validateStartEndFlags('123abc', '200')).toThrowErrorMatchingSnapshot()
    })

    it('throws error when end is invalid', () => {
      expect(() => validateStartEndFlags('100', '234abc')).toThrowErrorMatchingSnapshot()
    })

    it('throws error when start is not before the end time', () => {
      expect(() => validateStartEndFlags('200', '100')).toThrowErrorMatchingSnapshot()
    })

    it('sets end time to current time if end time is too large', () => {
      const now = Date.now()
      const res = validateStartEndFlags('0', '9999999999999')
      expect(res).not.toBeUndefined()
      const [start, end] = res
      expect(start).toBe(0)
      expect(end).toBeLessThan(9999999999999)
      expect(end).toStrictEqual(now)
    })
  })
})
