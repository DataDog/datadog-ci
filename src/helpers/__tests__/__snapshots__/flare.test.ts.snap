// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flare getEndpointUrl should not throw error if the site is invalid and DD_CI_BYPASS_SITE_VALIDATION is set 1`] = `"https://datad0ge.com/api/ui/support/serverless/flare"`;

exports[`flare getEndpointUrl should return correct endpoint url 1`] = `"https://app.datadoghq.com/api/ui/support/serverless/flare"`;

exports[`flare getEndpointUrl should throw error if the site is invalid 1`] = `"Invalid site: datad0ge.com. Must be one of: datadoghq.com, datadoghq.eu, us3.datadoghq.com, us5.datadoghq.com, ap1.datadoghq.com, ddog-gov.com"`;

exports[`flare getEndpointUrl should use DEFAULT_DD_SITE if CI_SITE_ENV_VAR and SITE_ENV_VAR are not set 1`] = `"https://app.datadoghq.com/api/ui/support/serverless/flare"`;

exports[`flare getEndpointUrl should use SITE_ENV_VAR if CI_SITE_ENV_VAR is not set 1`] = `"https://us3.datadoghq.com/api/ui/support/serverless/flare"`;

exports[`flare validateFilePath throws an error when the file does not exist 1`] = `
"[Error] File path '/not-exists' not found. Please try again.
"
`;

exports[`flare validateFilePath throws an error when the file has already been added 1`] = `
"[!] File '/added' has already been added.
"
`;

exports[`flare validateStartEndFlags throws error when end is invalid 1`] = `"End time must be a time in milliseconds since Unix Epoch. '234abc' is not a number."`;

exports[`flare validateStartEndFlags throws error when end is specified but start is not specified 1`] = `"Start time is required when end time is specified. [--start]"`;

exports[`flare validateStartEndFlags throws error when start is invalid 1`] = `"Start time must be a time in milliseconds since Unix Epoch. '123abc' is not a number."`;

exports[`flare validateStartEndFlags throws error when start is not before the end time 1`] = `"Start time must be before end time."`;

exports[`flare validateStartEndFlags throws error when start is specified but end is not specified 1`] = `"End time is required when start time is specified. [--end]"`;
